core/src/main/java/org/apache/calcite/rex/RexTableInputRef.java

```diff
@@ -46,10 +47,23 @@ public class RexTableInputRef extends RexInputRef {
 
   private final RelTableRef tableRef;
 
-  private RexTableInputRef(RelTableRef tableRef, int index, RelDataType type) {
+  /**
+   * It's a tag, which ref is wrapped by out join. And it always emits nullable = true.
+   * Input ref will be true, if it exists in right's input of left-join
+   * or left's input of right-join are true
+   */
+  private final boolean wrapJoinNullable;
+
+  private RexTableInputRef(RelTableRef tableRef, int index, RelDataType type,
+      boolean wrapJoinNullable) {
     super(index, type);
+    if (wrapJoinNullable) {
+      // Nullable of type must be true, if this input ref is wrapped by left/right join.
+      assert type.isNullable();
+    }
     this.tableRef = tableRef;
```

Comment:
Use ```assert wrapJoinNullable && type.isNulable()```.

Reply:
Yes, It looks more concise
But, Code should be `assert !forceNullable || type.isNullable();`

Reply:
If possible, try also to add an error message to newly added assertions

Reply:
@asolimando 
Yes, you are right. It's a good habit.